src/components/core/PivotTable.tsx

```diff
@@ -154,15 +161,92 @@ export const getTreeLeaves = (tree: any, getChildren = (node: any) => node && no
+    invariant(fieldId, `could not determine field type from ${colId}`);
+
+    const { attributeHeaders, measureHeaderItems } = assortDimensionHeaders(dimensions);
+
+    if (fieldType === 'a') {
+        for (const header of attributeHeaders) {
+            if (getIdsFromUri(header.attributeHeader.uri)[0] === fieldId) {
+                return {
+                    attributeSortItem: {
+                        direction,
+                        attributeIdentifier: header.attributeHeader.localIdentifier
+                    }
+                };
+            }
+        }
+    } else if (fieldType === 'm') {
+        const headerItem = measureHeaderItems[parseInt(fieldId, 10)];
+        const attributeLocators = fields.slice(0, -1).map((attributeField: string) => {
+            const [, attributeId, attributeValueId] = attributeField.match(/a_(\d*)_(\d*)/);
+            const attributeHeaderMatch = attributeHeaders.find((attributeHeader: Execution.IAttributeHeader) => {
+                return getIdsFromUri(attributeHeader.attributeHeader.formOf.uri)[0] === attributeId;
+            });
+            invariant(attributeHeaderMatch, `Could not find matching attribute header to field ${attributeField}`);
+            return {
+                attributeLocatorItem: {
+                    attributeIdentifier: attributeHeaderMatch.attributeHeader.localIdentifier,
+                    element: `${attributeHeaderMatch.attributeHeader.formOf.uri}/elements?id=${attributeValueId}`
+                }
+            };
+        });
+        return {
+            measureSortItem: {
+                direction,
+                // Type: LocatorItem[]
+                locators: [
+                    ...attributeLocators,
+                    {
+                        measureLocatorItem: {
+                            measureIdentifier: headerItem.measureHeaderItem.localIdentifier
+                        }
+                    }
+                ]
+            }
+        };
+    }
+    invariant(false, `could not find header matching ${colId}`);
+};
+
 export const getGridDataSource = (
     resultSpec: AFM.IResultSpec,
     getPage: IGetPage,
+    getExecution: () => Execution.IExecutionResponses,
     onSuccess: (execution: Execution.IExecutionResponses, columnDefs: IGridHeader[]) => void,
     columnDefOptions: IColumnDefOptions = {}
-) => ({
-    getRows: ({ startRow, endRow, successCallback }: IGetRowsParams) => {
+): IDatasource => ({
+    getRows: ({ startRow, endRow, successCallback, sortModel }: IGetRowsParams) => {
+        const execution = getExecution();
+        // If execution is null, this means this is a fresh dataSource and we should ignore current sortModel
+        const resultSpecWithSorting = (sortModel.length > 0 && execution)
+            ? {
+                ...resultSpec,
+                // override sorting based on sortModel
+                sorts: sortModel.map((sortModelItem: any) => {
```

Comment:
move to separate function eg. getSortsFromModel

Reply:
done